src/echo-common/src/smart-cache.ts

```diff
@@ -0,0 +1,186 @@
+
+export type SmartCacheLoadConfig = {
+  key: string,
+  mode?: SmartCacheLoadMode
+}
+
+export type SmartCacheBaseEvent = {
+  key: string,
+  timestampMs: number,
+}
+
+export type SmartCacheResultEvent<T> = SmartCacheBaseEvent & {
+  type: "result",
+  source: "cache-local" | "cache-memory" | "live",
+  result: T | null | undefined,
+}
+
+export type SmartCacheLoadingEvent = SmartCacheBaseEvent & {
+  type: "loading",
+}
+
+export type SmartCacheErrorEvent = SmartCacheBaseEvent & {
+  type: "error",
+  error: any
+}
+
+export type SmartCacheRateLimitEvent = SmartCacheBaseEvent & {
+  type: "rate-limit",
+  limitExpiresMs: number
+}
+
+export type SmartCacheStatusEvent = SmartCacheLoadingEvent | SmartCacheErrorEvent | SmartCacheRateLimitEvent
+
+export type SmartCacheEvent<T> = SmartCacheStatusEvent | SmartCacheResultEvent<T>
+
+export type SmartCacheStore<T> = {
+  lastResultEvent: SmartCacheResultEvent<T> | undefined,
+  lastStatusEvent: SmartCacheStatusEvent | undefined,
+}
+
+export type SmartCacheJob = {
+  config: SmartCacheLoadConfig
+}
+
+export class SmartCache<T> {
+  private jobQueue$ = new Subject<SmartCacheJob>()
+  private events$ = new Subject<SmartCacheEvent<T>>()
+
+  public memoryCache$ = new BehaviorSubject<{ [key: string]: SmartCacheStore<T> }>({})
+  private localCacheChecked: { [key: string]: boolean } = {}
+  private lastFireTimestampMs: number = 0
+
+  constructor(
+    private dir: EchoDirService,
+    loadFun: (key: string) => Observable<T | null>
+  ) {
+    this.events$.subscribe((event) => {
+      if (!(event.type === "result" && event.source === "cache-memory")) {
+        const currentStore = this.memoryCache$.value[event.key] ?? {}
+        const nextStore = { ...currentStore }
+
+        if (event.type === "result") {
+          nextStore.lastResultEvent = { ...event, source: "cache-memory" }
+          if (event.source === "live" && event.timestampMs > (currentStore.lastStatusEvent?.timestampMs ?? 0)) {
+            delete nextStore.lastStatusEvent
+          }
+        } else {
+          nextStore.lastStatusEvent = event
+        }
+
+        this.memoryCache$.next({ ...this.memoryCache$.value, [event.key]: nextStore })
+      }
+    })
+
+    this.jobQueue$
+      .pipe(
+        map((job) => this.loadFromLocalIfValid(job)),
+        filterNullish(),
+        groupBy((job) => job.config.key),
+        mergeMap((group) => group.pipe(throttleTime(10_000))),
+        // The throttleTime is for dedupuing multiple requests made while requests are firing
+        // Really the bypass needs to 99% of the time be in the isValid
+        map((job) => this.loadFromLocalIfValid(job)),
+        filterNullish(),
+        concatMap((job) => of(job)
+          .pipe(
+            delay(this.calulateRatelimitDelay()),
+            concatMap((job) => {
+              console.log("firing", job)
+              this.lastFireTimestampMs = Date.now()
+              return loadFun(job.config.key)
```

Comment:
The actual `loadFun` is still behind the `throttleTime(10_000)`, how to bypass this throttle?

This would cause the `useHook` -> `loadNow: () => load().subscribe()` to wait 10sec instead to load the data immediately. And outsite of the react hooks, we have the same problems.

Reply:
I think the solution is to only cache the actual load function. The `throttleTime` could be an additional `obserable` for subscribtions.
THis way we can trigger the `loadFun` immediately or we want a `throttleTime` subsubscribtion I still don't know what the advantage has this logic.